File: Rebex Components R6.11 - Samples/net-5.0/Imap/ImapBrowser_VB/ImapFolderSelector.vb
'  
'   Rebex Sample Code License
' 
'   Copyright 2023, REBEX CR s.r.o.
'   All rights reserved.
'   https://www.rebex.net/
' 
'   Permission to use, copy, modify, and/or distribute this software for any
'   purpose with or without fee is hereby granted.
' 
'   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
'   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
'   OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
'   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
'   HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
'   WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
'   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
'   OTHER DEALINGS IN THE SOFTWARE.
' 


Imports System
Imports System.Collections
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Windows.Forms
Imports Rebex.Net


''' <summary>
''' Summary description for ImapFolderSelector.
''' </summary>

Public Class ImapFolderSelector
    Inherits System.Windows.Forms.UserControl

    Private WithEvents treeView As System.Windows.Forms.TreeView

    Private _imap As Imap
    Private _lastSelected As TreeNode
    Private _initialized As Boolean

    Public ReadOnly Property IsInitialized() As Boolean
        Get
            Return _initialized
        End Get
    End Property

    Public ReadOnly Property SelectedFolder() As String
        Get
            Dim node As TreeNode = treeView.SelectedNode
            If node Is Nothing Then
                Return Nothing
            End If
            Dim tag As Object() = CType(node.Tag, Object())
            Dim folder As ImapFolder = CType(tag(0), ImapFolder)

            If Not folder.IsSelectable Then
                Return Nothing
            End If
            Return folder.Name
        End Get
    End Property

    Public Event SelectedFolderChanged As EventHandler


    ''' <summary>
    ''' Sets the internal IMAP object.
    ''' </summary>
    Public Sub SetClient(ByVal imap As Imap)
        _imap = imap
        If _initialized OrElse imap Is Nothing Then
            Return
        End If
        GetFolderList(treeView.Nodes, "", Nothing)

        _initialized = True

        Dim node As TreeNode
        For Each node In treeView.Nodes
            If String.Compare(node.Text, "Inbox", True) = 0 Then
                treeView.SelectedNode = node
                Exit For
            End If
        Next node
    End Sub       'SetClient


    Private Sub GetFolderList(ByVal nodes As TreeNodeCollection, ByVal name As String, ByVal delimiter As String)
        Cursor.Current = Cursors.WaitCursor
        Dim folders As ImapFolderCollection = _imap.GetFolderList(name, ImapFolderListMode.All, False)
        Cursor.Current = Cursors.Default
        Dim folder As ImapFolder
        For Each folder In folders
            Dim folderName As String = folder.Name
            If Not (delimiter Is Nothing) Then
                Dim p As Integer = folderName.LastIndexOf(delimiter)
                If p >= 0 Then
                    folderName = folderName.Substring((p + delimiter.Length))
                End If
            End If
            Dim node As New TreeNode(folderName)
            node.Tag = New Object() {folder, folder.CanContainInferiors}
            nodes.Add(node)
        Next folder
    End Sub       'GetFolderList

    ''' <summary> 
    ''' Required designer variable.
    ''' </summary>
    Private components As System.ComponentModel.Container = Nothing


    Public Sub New()
        ' This call is required by the Windows.Forms Form Designer.
        InitializeComponent()
    End Sub       'New


    ''' <summary> 
    ''' Clean up any resources being used.
    ''' </summary>
    Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
        If disposing Then
            If Not (components Is Nothing) Then
                components.Dispose()
            End If
        End If
        MyBase.Dispose(disposing)
    End Sub       'Dispose

#Region "Component Designer generated code"

    ''' <summary> 
    ''' Required method for Designer support - do not modify 
    ''' the contents of this method with the code editor.
    ''' </summary>
    Private Sub InitializeComponent()
        Me.treeView = New System.Windows.Forms.TreeView
        Me.SuspendLayout()
        ' 
        ' treeView
        ' 
        Me.treeView.Dock = System.Windows.Forms.DockStyle.Fill
        Me.treeView.ImageIndex = -1
        Me.treeView.Location = New System.Drawing.Point(0, 0)
        Me.treeView.Name = "treeView"
        Me.treeView.SelectedImageIndex = -1
        Me.treeView.Size = New System.Drawing.Size(150, 150)
        Me.treeView.TabIndex = 0
        ' 
        ' ImapFolderSelector
        ' 
        Me.Controls.Add(treeView)
        Me.Name = "ImapFolderSelector"
        Me.ResumeLayout(False)
    End Sub       'InitializeComponent 
#End Region


    Private Sub treeView_MouseDown(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles treeView.MouseDown
        If Not Enabled Then
            Return
        End If
        Dim sel As TreeNode = treeView.GetNodeAt(e.X, e.Y)
        If Not (sel Is Nothing) AndAlso Not treeView.SelectedNode Is sel Then
            treeView.SelectedNode = sel
        End If
    End Sub       'treeView_MouseDown

    Public Sub RefreshFolder()
        Dim node As TreeNode = treeView.SelectedNode
        Dim tag As Object() = CType(node.Tag, Object())
        Dim folder As ImapFolder = CType(tag(0), ImapFolder)
        Dim expandable As Boolean = CBool(tag(1))
        If expandable AndAlso Not (_imap Is Nothing) Then
            GetFolderList(node.Nodes, folder.Name, folder.Delimiter)
            node.Tag = New Object() {folder, False}
        End If
    End Sub       'RefreshFolder


    Private Sub treeView_AfterSelect(ByVal sender As Object, ByVal e As System.Windows.Forms.TreeViewEventArgs) Handles treeView.AfterSelect
        If Not Enabled Then
            treeView.SelectedNode = _lastSelected
            Return
        End If

        Dim node As TreeNode = treeView.SelectedNode
        If node Is _lastSelected Then
            Return
        End If
        _lastSelected = node
        If node Is Nothing Then
            Return
        End If

        RaiseEvent SelectedFolderChanged(Me, New EventArgs)
    End Sub       'treeView_AfterSelect
End Class   'ImapFolderSelector 'Rebex.Samples.ImapBrowser

File: Rebex Components R6.11 - Samples/net-4.6/Imap/ImapBrowser_CS/ImapFolderSelector.cs
//  
//   Rebex Sample Code License
// 
//   Copyright 2023, REBEX CR s.r.o.
//   All rights reserved.
//   https://www.rebex.net/
// 
//   Permission to use, copy, modify, and/or distribute this software for any
//   purpose with or without fee is hereby granted.
// 
//   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
//   OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//   HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
//   WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
//   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
//   OTHER DEALINGS IN THE SOFTWARE.
// 

using System;
using System.Collections;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Windows.Forms;
using Rebex.Net;

namespace Rebex.Samples
{
    /// <summary>
    /// Control which represents IMAP folder tree.
    /// </summary>
    public class ImapFolderSelector : System.Windows.Forms.UserControl
    {
        private System.Windows.Forms.TreeView treeView;

        private Imap _imap;
        private TreeNode _lastSelected;
        private bool _initialized;

        public bool IsInitialized
        {
            get{ return _initialized; }
        }

        public string SelectedFolder
        {
            get
            {
                TreeNode node = treeView.SelectedNode;
                if (node == null)
                    return null;

                object[] tag = (object[])node.Tag;
                ImapFolder folder = (ImapFolder)tag[0];

                if (!folder.IsSelectable)
                    return null;

                return folder.Name;
            }
        }

        public event EventHandler SelectedFolderChanged;

        /// <summary>
        /// Sets the internal IMAP object.
        /// </summary>
        public void SetClient(Imap imap)
        {
            _imap = imap;
            if (_initialized || imap == null)
                return;

            GetFolderList(treeView.Nodes, "", null);
            
            _initialized = true;

            foreach (TreeNode node in treeView.Nodes)
            {
                if (string.Compare(node.Text, "Inbox", true) == 0)
                {
                    treeView.SelectedNode = node;
                    break;
                }
            }			
        }

        private void GetFolderList(TreeNodeCollection nodes, string name, string delimiter)
        {
            Cursor.Current = Cursors.WaitCursor;
            ImapFolderCollection folders = _imap.GetFolderList(name, ImapFolderListMode.All, false);
            Cursor.Current = Cursors.Default;
            foreach (ImapFolder folder in folders)
            {
                string folderName = folder.Name;
                if (delimiter != null)
                {
                    int p = folderName.LastIndexOf(delimiter);
                    if (p >= 0)
                        folderName = folderName.Substring(p + delimiter.Length);
                }

                TreeNode node = new TreeNode(folderName);
                node.Tag = new object[]{folder, folder.CanContainInferiors};
                nodes.Add(node);
            }
        }

        /// <summary> 
        /// Required designer variable.
        /// </summary>
        private System.ComponentModel.Container components = null;

        public ImapFolderSelector()
        {
            // This call is required by the Windows.Forms Form Designer.
            InitializeComponent();

        }

        /// <summary> 
        /// Clean up any resources being used.
        /// </summary>
        protected override void Dispose( bool disposing )
        {
            if( disposing )
            {
                if(components != null)
                {
                    components.Dispose();
                }
            }
            base.Dispose( disposing );
        }

        #region Component Designer generated code
        /// <summary> 
        /// Required method for Designer support - do not modify 
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            this.treeView = new System.Windows.Forms.TreeView();
            this.SuspendLayout();
            // 
            // treeView
            // 
            this.treeView.Dock = System.Windows.Forms.DockStyle.Fill;
            this.treeView.ImageIndex = -1;
            this.treeView.Location = new System.Drawing.Point(0, 0);
            this.treeView.Name = "treeView";
            this.treeView.SelectedImageIndex = -1;
            this.treeView.Size = new System.Drawing.Size(150, 150);
            this.treeView.TabIndex = 0;
            this.treeView.MouseDown += new System.Windows.Forms.MouseEventHandler(this.treeView_MouseDown);
            this.treeView.AfterSelect += new System.Windows.Forms.TreeViewEventHandler(this.treeView_AfterSelect);
            // 
            // ImapFolderSelector
            // 
            this.Controls.Add(this.treeView);
            this.Name = "ImapFolderSelector";
            this.ResumeLayout(false);

        }
        #endregion

        private void treeView_MouseDown(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            if (!Enabled)
                return;

            TreeNode sel = treeView.GetNodeAt(e.X, e.Y);
            if (sel != null && treeView.SelectedNode != sel)
                treeView.SelectedNode = sel;
        }

        public void RefreshFolder()
        {
            TreeNode node = treeView.SelectedNode;
            object[] tag = (object[])node.Tag;
            ImapFolder folder = (ImapFolder)tag[0];
            bool expandable = (bool)tag[1];
            if (expandable && _imap != null)
            {
                GetFolderList(node.Nodes, folder.Name, folder.Delimiter);
                node.Tag = new object[] {folder, false};
            }
        }

        private void treeView_AfterSelect(object sender, System.Windows.Forms.TreeViewEventArgs e)
        {
            if (!Enabled)
            {
                treeView.SelectedNode = _lastSelected;
                return;
            }

            TreeNode node = treeView.SelectedNode;
            if (node == _lastSelected)
                return;

            _lastSelected = node;
            if (node == null)
                return;

            if (SelectedFolderChanged != null)
                SelectedFolderChanged(this, EventArgs.Empty);
        }
    }
}
